### PR TITLE
AI map

### PR DESCRIPTION
**Pull request by TobleroneSwordfish**
[FEATURE] [FREEZE]

## About the PR 
Integrates the goonhub.com interactive maps into the game as a map the AI can double click on to teleport its eye around. The map is opened from a verb "Open AI map" and can then be resized, zoomed etc.
This could also be made into a ghost or admin feature as well as, or instead of if you think giving this to the AI will make them a little _too_ omniscient.

map initial size

I haven't found any exploits with it so far, and it should be reasonably robust against them as it calls back to specifically the AI eye class, but I'm still a little paranoid that this could somehow introduce an arbitrary teleport exploit.
As for browser compatibility, it works on IE11 and should work on any browser the goonhub.com map works with. (Not sure what browsers goonstation/byond support)
The only known issue so far is that scrolling scrolls the map rather than zooming it. None of the usual JS/CSS tricks seem to fix this and it's difficult to diagnose as the scroll capture code for the map is in a minified JS file. Maybe someone who knows more about Byond window quirks can offer insight.
Performance impacts server-side should be minor as it's just serving a small HTML page. Client-side the map can take a few seconds to load the first time it's opened, but seems to be cached fine after that so re-opening it doesn't download the entire map again.
The code is mostly just a lot of HTML dumped directly into a string in ai-camera.dm, which makes me a little queasy but that's how QM terminals are implemented so I built off that.

Even better with multiple monitors
AI setup

## Why's this needed? 
Moving your camera around as AI can feel sluggish and clunky when you want to head to a specific spot rather than person. Viewports help with this, but you can only open so many.
A minimap is the only thing missing to complete my AI RTS fantasy.

## Changelog

```changelog
(u)LeahTheTech
(*)Added a map for the AI to move around with. Open with the "Open AI map" verb and double click to teleport.
```

### CODE REVIEW

**Comment by Arthur-Holiday** [comment]
seems fun. those maps don't get updated very often though, changes since their last updates probably aren't big enough to matter, but it is something I wouldn't rely on.

Does loading the maps take very long?

**Comment by TobleroneSwordfish** [comment]
> seems fun. those maps don't get updated very often though, changes since their last updates probably aren't big enough to matter, but it is something I wouldn't rely on.

Yeah, the actually robust way to do this would be to make a script to generate the map images from the map files directly (probably how the website maps were originally made?) and run that every time someone merges a map update. For now though they're easily good enough for teleporting between departments.

File: code/mob/dead/ai-camera.dm [hunk 1]
```diff
@@ -437,6 +445,97 @@
...
+		//default to cogmap if somehow the map settings go missing
+		//currently only Ozymandias and Icarus are unsupported
+		var/map_name = "cogmap"
+		if (map_settings)
+			var/position = findtext(map_settings.goonhub_map, "/maps/")
+			if (position > 0)
+				map_name = copytext(map_settings.goonhub_map, position + length("/maps/"))
+		//html dumped into source files fuck yeah
+		var/HTML = {"
+			<!DOCTYPE html>
```
Comment (ZeWaka) [inline code/mob/dead/ai-camera.dm:462]: yeah no

please make this into a tgui interface - it even have a special byond ui component to help with this

**Review by ZeWaka — CHANGES_REQUESTED** [review]

**Comment by ZeWaka** [comment]
> Yeah, the actually robust way to do this would be to make a script to generate the map images from the map files directly (probably how the website maps were originally made?) and run that every time someone merges a map update. For now though they're easily good enough for teleporting between departments.

They're made using the Map-World command at runtime.

File: code/mob/dead/ai-camera.dm (hunk 1 above)
Comment (TobleroneSwordfish) [reply code/mob/dead/ai-camera.dm:462]: Okay, cool. I am entirely on board with using modern UI frameworks. I'll start porting it now.

Comment (TobleroneSwordfish) [reply code/mob/dead/ai-camera.dm:462]: Just to clarify, do you mean make a tgui interface that that renders the website map, or one that uses a byond map component to render the map itself?

Comment (ZeWaka) [reply code/mob/dead/ai-camera.dm:462]: Either - but seeing as the former is what you did here, then the former.
Thinking on this again, you probably don't need the byondui component to catch the clicks or anything.

**Comment by ThatFiveGuys** [comment]
The ability to get coords from this map would be wonderful too.

**Pull request closed without merging**